### PR TITLE
Add auth2.0 RFC

### PR DESCRIPTION
**Pull request by donaldgray**

### CODE REVIEW

**Review by griffri — APPROVED** [review]

**Review by JackLewis-digirati — APPROVED** [review]

**Review by tomcrane — APPROVED** [review]

File: docs/rfcs/012-auth-service.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+AuthService creates a UserSession and related AccessToken, which is returned to the client as part of HTML payload. Provided `messageId` is echoed back to user.
+
+### [Probe Service](https://iiif.io/api/auth/2.0/#probe-service)
+
+> The probe service is used by the client to understand whether the user has access to the access-controlled resource for which the probe service is declared. The client sends the token obtained from the corresponding access token service to the probe service.
+
+Proposed endpoints:
+* Orchestrator `GET /auth/v2/probe/{asset-id}` (e.g. `GET /auth/v2/probe/10/20/foo`)
+  * Only the orchestrator knows the relationship between an asset + roles - the AuthService couldn't answer this on it's own.
+  * Responses can be shortcut from here - for example if the image doesn't require auth, or an access token hasn't been passed via `Authorization` header.
```
[the remark this answers is not in the document]
Comment (donaldgray) [reply docs/rfcs/012-auth-service.md:95]: Note added

File: docs/rfcs/012-auth-service.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+### Authentication
+
+For the CRUD operations, the new AuthService will need to validate that specified API keys are correct.
+
+For now, while using API-keys, we can delegate auth to DLCS - this is similar to how Composite-Handler will work.
+
+## Questions
+
+* When Orchestrator calls the ProbeService - do we want to pass roles in a header, rather than query parameter? The former wouldn't be logged. Is Image:Role relationship sensitive?
```
[the remark this answers is not in the document]
Comment (donaldgray) [reply docs/rfcs/012-auth-service.md:161]: I don't think there'd be any perf loss - passing as query param is certainly simpler. I agree that I don't think `Image:Role` is sensitive - the rolename in itself is just a URI so doesn't give anything away. I'll remove this question.

File: docs/rfcs/012-auth-service.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+* Should the `/probe/` endpoint in AuthService be locked down to Orchestrator only?
```
[the remark this answers is not in the document]
Comment (donaldgray) [reply docs/rfcs/012-auth-service.md:162]: Possibly - I'll update notes that it should be available internally only.

File: docs/rfcs/012-auth-service.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+* How do we handle the different paths for the DLCS as whole?
+  * Could this be an unnecessary complication?
+  * Using `/auth/v2/` prefix should allow us easily identify requests and direct accordingly.
+* Do Orchestrator and AuthService need be on the same domain? e.g. If Orchestrator is on `dlcs.example` and auth on `auth.dlcs.example` - would that lead to cookie issues? Thumbs service works in this way to may be a non-issue.
+* How do we know if the auth to be added to an image is 1 or 2 (or 3 etc in future)?
+  * The simplest method would be a customer specific setting - each customer would either be Auth1 or 2, not both.
+  * An alternative would be to have different roles - either something in the identifier specifies which Auth version (e.g. `http://dlcs.io/role/{version}/name`) - or some other source that we can lookup to determine which AuthService role X should use.
```
[the remark this answers is not in the document]
Comment (donaldgray) [reply docs/rfcs/012-auth-service.md:169]: The only wrinkle in this is when serving `/iiif-img/` requests - everything else (`AccessTokenService` etc would be different) but when working out "can this request view asset" we'd need to know which auth service to use.

For now we only have 1 + 2 so, as discussed earlier this week, we can use different cookie names for the different auth services. If a user has multiple cookies then we may need to check multiple auth services to get an answer (first `200` result wins).

**Pull request merged**